### PR TITLE
build: clean should delete out dir

### PR DESCRIPTION
**Pull request by lws-team**
'./out' dir is generated by the build action.

As well as .o, it also contains generated deps.

If you change your compiler, or optee version, the deps become invalid and
break the build.

It's the job of `make clean` to blow all this kind of thing away.

Signed-off-by: Andy Green <andy@warmcat.com>

### CODE REVIEW

**Comment by vchong** [comment]
Thanks Andy!

The maintainers don't allow use of `rm -rf` in Makefiles so please adjust accordingly.
Re: https://github.com/OP-TEE/build/pull/89#pullrequestreview-13963872 (See 'Show outdated')

HTH

**Comment by lws-team** [comment]
@vchong  Are you sure you understand the objection you linked to? ^^

```
Regardless what, don't over engineer this, the end goal is to delete a folder. As long as we avoid doing rm -rf $(VARIABLE) (i.e. a single variable) or rm -rf $(FOO)/mypath/$(BAR) 
```

I am not doing `rm -rf $(...)`.  I am doing `rm -rf ./out` which along with everything in it is a generated dir that will be regenerated next build.

**Comment by vchong** [comment]
@lws-team Perhaps this is still a gray area. I face similar objection(?) some time ago (https://github.com/OP-TEE/build/pull/49) ^^ but also see some uses [here](https://github.com/OP-TEE/build/search?utf8=%E2%9C%93&q=rm+-rf). Maybe it's ok now based on Joakim's latest comment above and also `I'm also fine with rm -rf $(VARIABLE)/some_odd_path_name`. Guess the maintainers will decide. :)

**Comment by lws-team** [comment]
@vchong: props for trying to fix it 9 months ago.

`make clean` deleting generated dirs is not a contentious operation.

If the maintainers have some opinion about it, let's see if we can discuss what they think is the correct way and commit it, and not leave such a simple thing broken another 9 months.

File: Makefile [hunk 1]
```diff
@@ -83,6 +83,7 @@ endif
 .PHONY: clean
 clean:
 	@$(cmd-echo-silent) '  CLEAN   .'
+	@${q}rm -rf out
```
Comment (jenswi-linaro) [inline Makefile]: The name of the out directory is `$(out-dir)`

**Comment by lws-team** [comment]
Mmm.  Well, there's also a very good reason I don't like projects that can't clean up after their build.

Do you think we should fix it so that it can clean up after itself, or leave it like it is?

If you have a specific idea to fix it, I can't do anything about a general indication someone might have said something at some point about the subject that you approve of.

Tell me what a fix should look like in your opinion.

**Comment by lws-team** [comment]
The PR is updated with ${out-dir} ... if there's something more to do, let me know what it is.

Since the build generates ${out-dir} can you help me understand what is 'sloppy' about removing exactly ${out-dir}?  It's be 'sloppy' removing anything else, wouldn't it?

**Comment by prime-zeng** [comment]
@lws-team  Have you considered the following case : 
Make O=$(out-dir) when compiling the optee-os, and the $(out-dir) is non-empty  ?

**Comment by jenswi-linaro** [comment]
A proper solution would in my mind be to update `$(cleanfiles)` so no files are left after `rm $(cleanfiles)`. After that step, all empty directories starting with the leafs should be removed.
If it's done this way it would be secure and not leave any build artifacts behind, wouldn't you agree?

**Comment by jenswi-linaro** [comment]
@jforissier, you're the best!

**Comment by jbech-linaro** [comment]
This line (Jeromes proposal):
```diff
+       ${q}if [ -d $(out-dir) ]; then rmdir `find $(out-dir) -depth -type d`; fi
```

`$(out-dir)` comes from the `O=` variable if O is defined, right? Set a side for a second that the question whether to clean all files in a folder vs a controlled clean (what is being proposed). The line above will delete all empty sub-folders to `O=`. So if I by mistake type `make O=$HOME`, then every single empty folder in my $HOME and it's sub directories will be deleted. In my opinion it's more safe to use `rm -rf $(VARIABLE)/some_unique_out_folder_name` than this proposal.

And regarding just deleting the "out" folder or doing a controlled delete, file by file, folder by folder, I have no problems just deleting the "out" folder. In the end it's a temporary working area for the Makefile, you're not really supposed to put other stuff there. In the end I don't care how it's done, just as long as it's empty after running `make clean`. So pick either the easy way (`rm -rf $(var)/unique_name`) or clean file by file, folder by folder, but please avoid doing tricks based on a single variable (or a using a variable at the end, since in that case on can goof up things by adding `../../` etc).

**Comment by lws-team** [comment]
Mmmm project build creates ./out/* entirely from generated files:

 - recreates ./out/ if missing

 - official path is Makefile has to track everything that is created in there or clean breaks ^^

 - clean target breaks for > 9 months: NGAF and ignores a patch that solves it.

 - the maintainers think the 'fix' is update the Makefile to current creation pattern

 - rm -rf is 'insecure' and 'sloppy'

Is that a fair summary? ^^

OK here's a suggestion:

 - generate the file in ${out-dir}/out and rm -rf ${out-dir}/out

Noooooooooo

**Comment by lws-team** [comment]
Do me a favour and just address the technical content.  It's your project that has the problem that needs fixing.

So far all I have been hearing is that `make clean` is sooo difficult it had to be left broken for >9 months.

It's such a security problem it might eat your home directory.

People in control are welcoming more complexity and unmaintainable tracking of generated files like it is the second coming.

How about we throw all that out along with whatever weird fetishization of rm -rf that has taken hold here and do what I said above... don't create the generated files in the O= dir.  Create it in a subdir of that, and delete the subdir during clean.

Correct me if I am wrong, but there are no technical objections left to that?

**Comment by lws-team** [comment]
I updated the PR with 2 patches

 - one removes cleanfiles (-46 ineffectual LOC)

 - the other adds an extra dir to O= or default O=, and makes it so `clean` rm -rf's the extra dir on top of O=

There are ramifications to this for guys who think they know where optee_os output product is (including uefi-tools).  And I think the TA "dev" makefiles should do the same, affecting that too. I wrote it up in the second patch changelog.

**Comment by jforissier** [comment]
OK, so both you Andy and @jbech-linaro are advocating the same option, which I think is acceptable and simpler than the current way. Let's do a quick poll: please vote for or against the below proposal using the reaction button (thumb up or down), and comment if you wish. Anyone sees a reason why this would not work? (projects using "make O=something" would need to be adjusted to pick the output files from the correct place but that's certainly not a big issue).

```diff
diff --git a/Makefile b/Makefile
index 33f434f..cdeddce 100644
--- a/Makefile
+++ b/Makefile
@@ -38,13 +38,11 @@ endif
 ARCH            ?= arm
 PLATFORM        ?= vexpress
 # Default value for PLATFORM_FLAVOR is set in plat-$(PLATFORM)/conf.mk
-O              ?= out/$(ARCH)-plat-$(PLATFORM)
+O              ?= out
 
 arch_$(ARCH)   := y
 
-ifneq ($O,)
-out-dir := $O
-endif
+out-dir := $O/$(ARCH)-plat-$(PLATFORM)
 
 ifneq ($V,1)
 q := @
@@ -82,8 +80,8 @@ endif
 
 .PHONY: clean
jerome@jfw540:~/work/hikey_optee/optee_os (fix-clean)$ vi Makefile 
jerome@jfw540:~/work/hikey_optee/optee_os (fix-clean)$ vi Makefile 
jerome@jfw540:~/work/hikey_optee/optee_os (fix-clean)$ gti diff
diff --git a/Makefile b/Makefile
index 33f434f..e0e10b7 100644
--- a/Makefile
+++ b/Makefile
@@ -38,13 +38,11 @@ endif
 ARCH            ?= arm
 PLATFORM        ?= vexpress
 # Default value for PLATFORM_FLAVOR is set in plat-$(PLATFORM)/conf.mk
-O              ?= out/$(ARCH)-plat-$(PLATFORM)
+O              ?= out
 
 arch_$(ARCH)   := y
 
-ifneq ($O,)
-out-dir := $O
-endif
+out-dir := $O/$(ARCH)-plat-$(PLATFORM)
 
 ifneq ($V,1)
 q := @
@@ -82,8 +80,9 @@ endif
 
 .PHONY: clean
 clean:
-       @$(cmd-echo-silent) '  CLEAN   .'
-       ${q}rm -f $(cleanfiles)
+       @$(cmd-echo-silent) '  CLEAN   $O'
+       ${q}rm -rf $O/$(ARCH)-plat-$(PLATFORM)
+       ${q}rmdir --ignore-fail-on-non-empty $O
 
 .PHONY: cscope
 cscope:
diff --git a/core/arch/arm/kernel/link.mk b/core/arch/arm/kernel/link.mk
index 2220fe6..bbe87f2 100644
--- a/core/arch/arm/kernel/link.mk
+++ b/core/arch/arm/kernel/link.mk
@@ -32,12 +32,10 @@ entries-unpaged += sem_cpu_sync
 entries-unpaged += generic_boot_get_handlers
 
 ldargs-all_objs := -i $(objs) $(link-ldadd) $(libgcccore)
-cleanfiles += $(link-out-dir)/all_objs.o
 $(link-out-dir)/all_objs.o: $(objs) $(libdeps) $(MAKEFILE_LIST)
        @$(cmd-echo-silent) '  LD      $@'
        $(q)$(LDcore) $(ldargs-all_objs) -o $@
 
-cleanfiles += $(link-out-dir)/unpaged_entries.txt
 $(link-out-dir)/unpaged_entries.txt: $(link-out-dir)/all_objs.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(NMcore) $< | \
@@ -49,26 +47,22 @@ objs-unpaged := \
        $(filter-out $(addprefix $(out-dir)/, $(objs-unpaged-rem)), $(objs))
 ldargs-unpaged = -i --gc-sections $(addprefix -u, $(entries-unpaged))
 ldargs-unpaged-objs := $(objs-unpaged) $(link-ldadd) $(libgcccore)
-cleanfiles += $(link-out-dir)/unpaged.o
 $(link-out-dir)/unpaged.o: $(link-out-dir)/unpaged_entries.txt
        @$(cmd-echo-silent) '  LD      $@'
        $(q)$(LDcore) $(ldargs-unpaged) \
                `cat $(link-out-dir)/unpaged_entries.txt` \
                $(ldargs-unpaged-objs) -o $@
 
-cleanfiles += $(link-out-dir)/text_unpaged.ld.S:
 $(link-out-dir)/text_unpaged.ld.S: $(link-out-dir)/unpaged.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(READELFcore) -a -W $< | ${AWK} -f ./scripts/gen_ld_text_sects.awk > $@
 
-cleanfiles += $(link-out-dir)/rodata_unpaged.ld.S:
 $(link-out-dir)/rodata_unpaged.ld.S: $(link-out-dir)/unpaged.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(READELFcore) -a -W $< | \
                ${AWK} -f ./scripts/gen_ld_rodata_sects.awk > $@
 
 
-cleanfiles += $(link-out-dir)/init_entries.txt
 $(link-out-dir)/init_entries.txt: $(link-out-dir)/all_objs.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(NMcore) $< | \
@@ -85,7 +79,6 @@ objs-init := \
 ldargs-init := -i --gc-sections $(addprefix -u, $(entries-init))
 
 ldargs-init-objs := $(objs-init) $(link-ldadd) $(libgcccore)
-cleanfiles += $(link-out-dir)/init.o
 $(link-out-dir)/init.o: $(link-out-dir)/init_entries.txt
        $(call gen-version-o)
        @$(cmd-echo-silent) '  LD      $@'
@@ -93,12 +86,10 @@ $(link-out-dir)/init.o: $(link-out-dir)/init_entries.txt
                `cat $(link-out-dir)/init_entries.txt` \
                $(ldargs-init-objs) -o $@
 
-cleanfiles += $(link-out-dir)/text_init.ld.S:
 $(link-out-dir)/text_init.ld.S: $(link-out-dir)/init.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(READELFcore) -a -W $< | ${AWK} -f ./scripts/gen_ld_text_sects.awk > $@
 
-cleanfiles += $(link-out-dir)/rodata_init.ld.S:
 $(link-out-dir)/rodata_init.ld.S: $(link-out-dir)/init.o
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(READELFcore) -a -W $< | \
@@ -111,7 +102,6 @@ link-script-extra-deps += $(link-out-dir)/rodata_unpaged.ld.S
 link-script-extra-deps += $(link-out-dir)/text_init.ld.S
 link-script-extra-deps += $(link-out-dir)/rodata_init.ld.S
 link-script-extra-deps += $(conf-file)
-cleanfiles += $(link-script-pp) $(link-script-dep)
 $(link-script-pp): $(link-script) $(link-script-extra-deps)
        @$(cmd-echo-silent) '  CPP     $@'
        @mkdir -p $(dir $@)
@@ -147,22 +137,17 @@ $(link-out-dir)/version.o:
        $(call gen-version-o)
 
 all: $(link-out-dir)/tee.elf
-cleanfiles += $(link-out-dir)/tee.elf $(link-out-dir)/tee.map
-cleanfiles += $(link-out-dir)/version.o
-cleanfiles += $(link-out-dir)/.buildcount
 $(link-out-dir)/tee.elf: $(objs) $(libdeps) $(link-script-pp)
        @$(cmd-echo-silent) '  LD      $@'
        $(q)$(LDcore) $(ldargs-tee.elf) -o $@
 
 all: $(link-out-dir)/tee.dmp
-cleanfiles += $(link-out-dir)/tee.dmp
 $(link-out-dir)/tee.dmp: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  OBJDUMP $@'
        $(q)$(OBJDUMPcore) -l -x -d $< > $@
 
 pageable_sections := .*_pageable
 init_sections := .*_init
-cleanfiles += $(link-out-dir)/tee-pager.bin
 $(link-out-dir)/tee-pager.bin: $(link-out-dir)/tee.elf \
                $(link-out-dir)/tee-data_end.txt
        @$(cmd-echo-silent) '  OBJCOPY $@'
@@ -172,7 +157,6 @@ $(link-out-dir)/tee-pager.bin: $(link-out-dir)/tee.elf \
                --pad-to `cat $(link-out-dir)/tee-data_end.txt` \
                $< $@
 
-cleanfiles += $(link-out-dir)/tee-pageable.bin
 $(link-out-dir)/tee-pageable.bin: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  OBJCOPY $@'
        $(q)$(OBJCOPYcore) -O binary \
@@ -180,32 +164,27 @@ $(link-out-dir)/tee-pageable.bin: $(link-out-dir)/tee.elf
                --only-section="$(pageable_sections)" \
                $< $@
 
-cleanfiles += $(link-out-dir)/tee-data_end.txt
 $(link-out-dir)/tee-data_end.txt: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        @echo -n 0x > $@
        $(q)$(NMcore) $< | grep __data_end | sed 's/ .*$$//' >> $@
 
-cleanfiles += $(link-out-dir)/tee-init_size.txt
 $(link-out-dir)/tee-init_size.txt: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        @echo -n 0x > $@
        $(q)$(NMcore) $< | grep __init_size | sed 's/ .*$$//' >> $@
 
-cleanfiles += $(link-out-dir)/tee-init_load_addr.txt
 $(link-out-dir)/tee-init_load_addr.txt: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        @echo -n 0x > $@
        $(q)$(NMcore) $< | grep ' _start' | sed 's/ .*$$//' >> $@
 
-cleanfiles += $(link-out-dir)/tee-init_mem_usage.txt
 $(link-out-dir)/tee-init_mem_usage.txt: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        @echo -n 0x > $@
        $(q)$(NMcore) $< | grep ' __init_mem_usage' | sed 's/ .*$$//' >> $@
 
 all: $(link-out-dir)/tee.bin
-cleanfiles += $(link-out-dir)/tee.bin
 $(link-out-dir)/tee.bin: $(link-out-dir)/tee-pager.bin \
                         $(link-out-dir)/tee-pageable.bin \
                         $(link-out-dir)/tee-init_size.txt \
@@ -225,12 +204,10 @@ $(link-out-dir)/tee.bin: $(link-out-dir)/tee-pager.bin \
 
 
 all: $(link-out-dir)/tee.symb_sizes
-cleanfiles += $(link-out-dir)/tee.symb_sizes
 $(link-out-dir)/tee.symb_sizes: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(NMcore) --print-size --reverse-sort --size-sort $< > $@
 
-cleanfiles += $(link-out-dir)/tee.mem_usage
 ifneq ($(filter mem_usage,$(MAKECMDGOALS)),)
 mem_usage: $(link-out-dir)/tee.mem_usage
 
diff --git a/core/arch/arm/plat-rcar/link.mk b/core/arch/arm/plat-rcar/link.mk
index 763298f..b16280d 100644
--- a/core/arch/arm/plat-rcar/link.mk
+++ b/core/arch/arm/plat-rcar/link.mk
@@ -1,7 +1,6 @@
 include core/arch/arm/kernel/link.mk
 
 all: $(link-out-dir)/tee.srec
-cleanfiles += $(link-out-dir)/tee.srec
 $(link-out-dir)/tee.srec: $(link-out-dir)/tee.elf
        @$(cmd-echo-silent) '  GEN     $@'
        $(q)$(OBJCOPYcore) -O srec $< $@
diff --git a/core/arch/arm/plat-sunxi/link.mk b/core/arch/arm/plat-sunxi/link.mk
index 2e289e6..5e220fb 100644
--- a/core/arch/arm/plat-sunxi/link.mk
+++ b/core/arch/arm/plat-sunxi/link.mk
@@ -8,10 +8,6 @@ AWK     = awk
 
 all: $(link-out-dir)/tee.elf $(link-out-dir)/tee.dmp $(link-out-dir)/tee.bin
 all: $(link-out-dir)/tee.symb_sizes
-cleanfiles += $(link-out-dir)/tee.elf $(link-out-dir)/tee.dmp $(link-out-dir)/tee.map
-cleanfiles += $(link-out-dir)/tee.bin
-cleanfiles += $(link-out-dir)/tee.symb_sizes
-cleanfiles += $(link-script-pp) $(link-script-dep)
 
 link-ldflags  = $(LDFLAGS)
 link-ldflags += -T $(link-script-pp) -Map=$(link-out-dir)/tee.map
diff --git a/core/core.mk b/core/core.mk
index 075bc01..9f641df 100644
--- a/core/core.mk
+++ b/core/core.mk
@@ -60,9 +60,6 @@ conf-file := $(out-dir)/include/generated/conf.h
 conf-mk-file := $(out-dir)/conf.mk
 $(conf-file): $(conf-mk-file)
 
-cleanfiles += $(conf-file)
-cleanfiles += $(conf-mk-file)
-
 $(conf-file): FORCE
        $(call check-conf-h)
 
diff --git a/mk/compile.mk b/mk/compile.mk
index d4fb632..3c60efe 100644
--- a/mk/compile.mk
+++ b/mk/compile.mk
@@ -7,7 +7,6 @@
 # Output
 #
 # set    objs
-# update  cleanfiles
 #
 # Generates explicit rules for all objs
 
@@ -65,8 +64,6 @@ comp-cmd-file-$2:= $$(dir $2).$$(notdir $2).cmd
 comp-sm-$2     := $(sm)
 comp-lib-$2    := $(libname)-$(sm)
 
-cleanfiles := $$(cleanfiles) $$(comp-dep-$2) $$(comp-cmd-file-$2) $2
-
 ifeq ($$(filter %.c,$1),$1)
 comp-q-$2 := CC
 comp-flags-$2 = $$(filter-out $$(CFLAGS_REMOVE) $$(cflags-remove) \
@@ -162,8 +159,6 @@ comp-dep-$3 := $$(dir $3)$$(notdir $3).d
 comp-cmd-file-$3:= $$(dir $3)$$(notdir $3).cmd
 comp-sm-$3     := $(sm)
 
-cleanfiles := $$(cleanfiles) $$(comp-dep-$3) $$(comp-cmd-file-$3) $3 $2
-
 comp-flags-$3 = $$(filter-out $$(CFLAGS_REMOVE) $$(cflags-remove) \
                              $$(cflags-remove-$$(comp-sm-$3)) \
                              $$(cflags-remove-$3), \
diff --git a/mk/lib.mk b/mk/lib.mk
index 87cf579..d00d272 100644
--- a/mk/lib.mk
+++ b/mk/lib.mk
@@ -7,7 +7,6 @@
 #
 # Output
 #
-# updated cleanfiles and
 # updated libfiles, libdirs, libnames and libdeps
 
 
@@ -16,7 +15,6 @@ include mk/subdir.mk
 include mk/compile.mk
 
 lib-libfile     = $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).a
-cleanfiles     := $(cleanfiles) $(lib-libfile)
 libfiles       := $(lib-libfile) $(libfiles)
 libdirs        := $(out-dir)/$(base-prefix)$(libdir) $(libdirs)
 libnames       := $(libname) $(libnames)
```

**Comment by lws-team** [comment]
The only differences I see are

 - if he built multiple arch builds, make clean should nuke them all.  The only safe way to get that effect is add a dir level.

 - if this is good for optee_os, it should be good for external "dev kit" TA builds.

Plus I think optee_client may have the same disease but that's a problem for another day

Otherwise for me, I don't mind either way.  I only have one arch.

**Comment by lws-team** [comment]
+ adding the dir level needs to align travis...

**Comment by jforissier** [comment]
Your patch doesn't work.
```
jerome@jfw540:~/work/hikey_optee/optee_os ((c9588cc...))$ make -s -j9
jerome@jfw540:~/work/hikey_optee/optee_os ((c9588cc...))$ make clean
  CLEAN   .
rm: cannot remove 'out/optee_os_out/arm-plat-vexpress/..': Is a directory
Makefile:86: recipe for target 'clean' failed
make: *** [clean] Error 1
```

> The only differences I see are
> 
>  - if he built multiple arch builds, make clean should nuke them all.  The only safe way to get that effect is add a dir level.

I'd rather clean only one arch actually. More flexible and more compatible with the current layout.

> 
>  - if this is good for optee_os, it should be good for external "dev kit" TA builds.
>
> Plus I think optee_client may have the same disease but that's a problem for another day

Exactly, let's handle one issue at a time.

**Comment by jenswi-linaro** [comment]
I'd prefer if we could keep the meaning of `O=`, that is not change the structure below as this is bound to force everyone to update their build scripts.

**Comment by lws-team** [comment]
Yeah... although it is more correct to nuke a sacrificial dir containing all build products, it will indeed make a lot of trouble to start doing it now.  Judging it by the 'is it going to be better than what we have' test we should just solve this using jforissier's patch.  I have no more time to spend on it and the two patches are pretty identical except the extra dir.

I'm happy to just have inspired the solution... please close this PR when jforissier's patch goes in.

It'd be good if someone can eyeball optee_client.  optee_test has different problems with make clean https://github.com/OP-TEE/optee_test/issues/162 ... I have a 'solution' posted there at the same level as rm -rf out, ie, scratches only my itch.  I duno what a good solution would look like, maybe clean stanzas in every sub.mk.

**Comment by jforissier** [comment]
If we don't want to change the structure under `$(O)` (backward compatibility), then we must rule out `rm -rf $(O)/some_dir` I'm afraid.
I have pushed a solution that hopefully addresses all the requirements, but is a bit complex unfortunately -- for those who don't speak gmake fluently, that is ;)
See https://github.com/OP-TEE/optee_os/pull/1275.

**Comment by jforissier** [comment]
Closing this PR, let's discuss the new one.

**Pull request closed without merging**